source/extensions/transport_sockets/s2a/tsi_socket.cc

```diff
@@ -0,0 +1,389 @@
+    raw_write_buffer_.move(*next_result->to_send_);
+  }
+
+  if (status == TSI_OK && handshaker_result != nullptr) {
+    tsi_peer peer;
+    // returns TSI_OK assuming there is no fatal error. Asserting OK.
+    status = tsi_handshaker_result_extract_peer(handshaker_result, &peer);
+    ASSERT(status == TSI_OK);
+    Cleanup peer_cleanup([&peer]() { tsi_peer_destruct(&peer); });
+    ENVOY_CONN_LOG(debug, "TSI: Handshake successful: peer properties: {}",
+                   callbacks_->connection(), peer.property_count);
+    for (size_t i = 0; i < peer.property_count; ++i) {
+      ENVOY_CONN_LOG(debug, "  {}: {}", callbacks_->connection(), peer.properties[i].name,
+                     std::string(peer.properties[i].value.data, peer.properties[i].value.length));
+    }
+    if (handshake_validator_) {
+      std::string err;
+      TsiInfo tsi_info;
+      const bool peer_validated = handshake_validator_(peer, tsi_info, err);
+      if (peer_validated) {
+        ENVOY_CONN_LOG(debug, "TSI: Handshake validation succeeded.", callbacks_->connection());
+      } else {
+        ENVOY_CONN_LOG(debug, "TSI: Handshake validation failed: {}", callbacks_->connection(),
+                       err);
+        return Network::PostIoAction::Close;
+      }
+      ProtobufWkt::Struct dynamic_metadata;
+      ProtobufWkt::Value val;
+      val.set_string_value(tsi_info.name_);
+      dynamic_metadata.mutable_fields()->insert({std::string("peer_identity"), val});
+      callbacks_->connection().streamInfo().setDynamicMetadata(
+          "envoy.transport_sockets.peer_information", dynamic_metadata);
+      ENVOY_CONN_LOG(debug, "TSI hanshake with peer: {}", callbacks_->connection(), tsi_info.name_);
+    } else {
+      ENVOY_CONN_LOG(debug, "TSI: Handshake validation skipped.", callbacks_->connection());
+    }
+
+    const unsigned char* unused_bytes;
+    size_t unused_byte_size;
+
+    // returns TSI_OK assuming there is no fatal error. Asserting OK.
+    status =
+        tsi_handshaker_result_get_unused_bytes(handshaker_result, &unused_bytes, &unused_byte_size);
+    ASSERT(status == TSI_OK);
+    if (unused_byte_size > 0) {
+      // All handshake data is consumed.
+      ASSERT(raw_read_buffer_.length() == 0);
+      raw_read_buffer_.prepend(
+          absl::string_view{reinterpret_cast<const char*>(unused_bytes), unused_byte_size});
+    }
+    ENVOY_CONN_LOG(debug, "TSI: Handshake successful: unused_bytes: {}", callbacks_->connection(),
+                   unused_byte_size);
+
+    // returns TSI_OK assuming there is no fatal error. Asserting OK.
+    tsi_zero_copy_grpc_protector* frame_protector;
+    status = tsi_handshaker_result_create_zero_copy_grpc_protector(
+        handshaker_result, &default_max_frame_size_, &frame_protector);
+    ASSERT(status == TSI_OK);
+
+    // TODO(tavishvaidya): Check the return value once fake TSI frame protector
+    // used in tsi_socket_test.cc implements the interface returning the max frame size.
```

Comment:
@dmbm0830 Can we actually implement this now? Since the changes are to the fake TSI frame protector, is there anything on the TSI interface side that prevents this change? 
@matthewstevenson88 FYI.

Reply:
I think we can do this, based on this recent PR: https://github.com/grpc/grpc/pull/25621